apps/server/src/modules/board/repo/board-do.repo.ts

```diff
@@ -0,0 +1,86 @@
+import { Utils } from '@mikro-orm/core';
+import { EntityManager } from '@mikro-orm/mongodb';
+import { Injectable } from '@nestjs/common';
+import { AnyBoardDo, BoardNode, EntityId } from '@shared/domain';
+import { BoardDoBuilder } from '@shared/domain/entity/boardnode/board-do.builder';
+import { BoardNodeBuilderImpl } from '@shared/domain/entity/boardnode/board-node-builder-impl';
+import { BoardNodeRepo } from './board-node.repo';
+
+@Injectable()
+export class BoardDoRepo {
+	constructor(private readonly em: EntityManager, private readonly boardNodeRepo: BoardNodeRepo) {}
+
+	async findById(id: EntityId): Promise<AnyBoardDo> {
+		const boardNode = await this.em.findOneOrFail(BoardNode, id);
+		const descendants = await this.boardNodeRepo.findDescendants(boardNode);
+		const domainObject = new BoardDoBuilder(descendants).buildDomainObject(boardNode);
+
+		return domainObject;
+	}
+
+	async findByIds(ids: EntityId[]): Promise<AnyBoardDo[]> {
+		const boardNodes = await this.em.find(BoardNode, { id: { $in: ids } });
+
+		const childrenMap = await this.boardNodeRepo.findChildrenOfMany(boardNodes);
+
+		const domainObjects = boardNodes.map((boardNode) => {
+			const children = childrenMap[boardNode.pathOfChildren];
+			const domainObject = new BoardDoBuilder(children).buildDomainObject(boardNode);
+			return domainObject;
+		});
+
+		return domainObjects;
+	}
+
+	async findParentOfId(childId: EntityId): Promise<AnyBoardDo> {
+		const boardNode = await this.em.findOneOrFail(BoardNode, childId);
+		if (boardNode.parentId) {
+			const parent = await this.em.findOneOrFail(BoardNode, boardNode.parentId);
+			const descendants = await this.boardNodeRepo.findDescendants(parent);
+			const domainObject = new BoardDoBuilder(descendants).buildDomainObject(parent);
+			return domainObject;
+		}
+		throw new Error('boardNode with given childId has no parent');
+	}
+
+	async save(domainObject: AnyBoardDo | AnyBoardDo[], parentId?: EntityId) {
+		const getParent = async (id?: EntityId): Promise<BoardNode | undefined> =>
+			id ? this.boardNodeRepo.findById(BoardNode, id) : undefined;
+
+		const domainObjects = Utils.asArray(domainObject);
+		const parentNode = await getParent(parentId);
+		const builder = new BoardNodeBuilderImpl(parentNode);
+		const boardNodes = builder.buildBoardNodes(domainObjects, parentNode?.id);
+		await this.boardNodeRepo.save(boardNodes);
+	}
+
+	async deleteChild<T extends AnyBoardDo>(parent: T, childId: EntityId): Promise<T> {
```

Comment:
I think the repo is maybe not the right place to do that. This should rather be done in the domain layer. We should manipulate the domain objects (list of children) in the service. The repo should then be used to persist those changes.

Reply:
manipulating and updating the list of children is one thing we need to do, another is the actual removing from the database. that cant be done by the domain layer (because it cant delete things without a repo), and a function that just deletes a single node doesnt make any sense because it will always lead to inconsistent data, deleting a node always has to lead to updates on children and siblings. But we can discuss if you have a better idea...